### ISSUE TITLE
`SelectionArea` does not select the text when using double and triple tap

### ISSUE THREAD

**Issue by nilsreichardt** [flutter/flutter#104552]
## Description
This is an extracted feature request of @justinmc comment:
> Double-click to select a word doesn't work.

https://github.com/flutter/flutter/pull/95226#pullrequestreview-976138900

On HTML pages you can use double tap for selecting one word and triple tap for selecting the complete text block becomes selected:

https://user-images.githubusercontent.com/24459435/170142449-23d7e515-5355-4370-be68-19c7d67de2e9.mov

But if try it in Flutter, nothing happens:

https://user-images.githubusercontent.com/24459435/170142471-1dbdd898-72b1-44d0-b97b-37d932af718c.mov

cc: @chunhtai

**Comment by letungcntt** [comment]
@nilsreichardt Can you add Shift + click to selection text ?

**Comment by nilsreichardt** [comment]
Here is also an example when using shift + click on an HTML website:

https://user-images.githubusercontent.com/24459435/170472540-03a653e6-11ac-4e41-80a8-dc0ba4c66a28.mov

@chunhtai Should I open a separate issue for this or just us this ticket?

**Comment by chunhtai** [comment]
@nilsreichardt Let's create a different one.

**Comment by chunhtai** [comment]
this looks like a mouse only gesture, I tested on touch screen, they can only select through long press

**Comment by letungcntt** [comment]
Any update this issue ?

**Comment by chunhtai** [comment]
currently wait for gesture refactoring https://github.com/flutter/flutter/pull/109573. Once that is done, selectable region can reuse the gestures from editable text

**Comment by AlexeyKatsuro** [comment]
Any news? Still hoping.

I would like to add that in the video example the author uses [google-translate browser extension](https://chrome.google.com/webstore/detail/google-translate/aapbdbdomjkkjkaonfhkkikfgjllcleb)
which reacts to the selection of text and can also be triggered by double-clicking, but unfortunately, the selected text from the flutter does not work with this useful extension

**Comment by nico-famedly** [comment]
@AlexeyKatsuro https://github.com/flutter/flutter/pull/119046 is probably related

**Comment by justinmc** [comment]
@Renzo-Olivares will your work on triple tap in EditableText be able to help out here?

**Comment by Renzo-Olivares** [comment]
@justinmc Not immediately, some work still needs to be done to migrate `SelectionArea` to use `TapAndDragGestureRecognizer`. Then double tap/triple tap have to re-implemented for `SelectionArea`. I can take a look after #119046 gets merged.

**Comment by alexrabin** [comment]
@Renzo-Olivares looks like #119046 has been merged. Have you had time to take a look at this?

**Comment by Renzo-Olivares** [comment]
Hello @alexrabin, thanks for your interest. I am currently working on this.

**Comment by Morpa** [comment]
Is this issue still open?

**Comment by Morpa** [comment]
> @Morpa Hello, yes this issue is still open.

Thank you for your response

**Comment by Renzo-Olivares** [comment]
This issue is still open. #124817 added support for double click to select a word and double click + drag to select word-by-word. Triple click support is still needed.

**Comment by Renzo-Olivares** [comment]
Hi all, this is still being worked on. I am currently fixing some edge cases with selectable order https://github.com/flutter/flutter/pull/136920 , which is a prerequisite for triple click.

**Comment by Unbreathable** [comment]
I've got some more suggestions how to improve the user experience with the ``SelectionArea``, this might not be a fit for this issue, but since it is about double click selection I thought I would put it here anyway: 

Double click selection of course works if you are clicking on the word itself, but most of the time the user actually also expects to select a word when clicking at the end of a line in empty space. It's hard for me to describe so here is a video of the behaviour for double/tripple click selection on a regular website inside of Firefox:

https://github.com/flutter/flutter/assets/70802809/6b84f037-eec4-4c89-9e4d-c3ce7cc64076

I think many users are just used to this and it would be good if it were included in Flutter as well. I'm not familiar enough with the Flutter codebase to do it myself unfortunately (maybe if I have the time?), but it's totally fine if you don't want to add this since you are working on this in your free time.

**Comment by Unbreathable** [comment]
Hello, I've just got a little suggestion to improve the user experience with ``SelectionArea`` a little more. Currently if you have text that is tightly wrapped you can't really select it properly when clicking outside of it. Let me show you what I mean, there isn't really sound in the video, but you'll just have to trust me that I'm double clicking to select.

https://github.com/flutter/flutter/assets/70802809/f61a55be-7d03-41a8-b78d-6ca1f35aee76

The darker box is the width the text can actually use, but since it is so tightly wrapped the text tries taking as little width as possible making the ``SelectionArea`` probably just not recognize that the user might be trying to select it. From what I'm guessing it's probably hard to implement this since the ``SelectionArea`` would somehow have to identify this area since the ``Text`` is not expanding to it and the ``Text`` suddenly expanding in a new Flutter release might cause some UI to break. Although it would be cool if this could work similar to the browser where you can click pretty much anywhere that's not a button and start selecting the text around it.

**Comment by ndnam198** [comment]
@Renzo-Olivares are you still working on this ticket, I see it has been quite a long time since the last update. I am wondering if the feature is ready to use yet? Thanks

**Comment by vergaraSC** [comment]
@Renzo-Olivares double tap in flutter web with Disable the browser's context menu, its possible?

**Comment by Renzo-Olivares** [comment]
Hi @ndnam198, this is still being worked on, I have been working on this along with some related work on selectable ordering. This should be ready in the next few weeks.

@vergaraSC double click should work when the browsers context menu is disabled. Is it not working for you, and if so on what platform are you on?

**Comment by vergaraSC** [comment]
@Renzo-Olivares 

I am using Flutter Web, in my initState i have this BrowserContextMenu.disableContextMenu(); (when is web).
And I use the following code with my custom contextMenuBuilder, it works fine, copy, selectAll, but when I try to double tap on a text it doesn't work, only when I right click on a text it opens my toolbar.

```
SelectionArea(
            key: selectionArea,
            focusNode: focusNode,
            magnifierConfiguration: const TextMagnifierConfiguration(),
            contextMenuBuilder: (context, editableTextState) {
              final keyContext = selectionArea.currentContext;
              final box = keyContext?.findRenderObject() as RenderBox;
              Offset position = box.localToGlobal(Offset.zero);

              return USFMContextMenu(
                anchor: editableTextState.contextMenuAnchors.primaryAnchor,
                selectionAreaPosition: position.dx - 4,
                children: [
                  TextSelectionToolbarTextButton(
                    padding: const EdgeInsets.all(4),
                    onPressed: () {
                      ContextMenuController.removeAny();
                      FocusScope.of(context).requestFocus(FocusNode());
                    },
                    child: const Icon(Icons.close),
                  ),
                  const USFMContextMenuDividerItem(),
                  TextSelectionToolbarTextButton(
                    padding: const EdgeInsets.all(4),
                    onPressed: () {
                      editableTextState.copySelection(
                        SelectionChangedCause.toolbar,
                      );
                    },
                    child: const Icon(Icons.copy),
                  ),
                  const USFMContextMenuDividerItem(),
                  TextSelectionToolbarTextButton(
                    padding: const EdgeInsets.all(4),
                    onPressed: () {
                      editableTextState.selectAll(
                        SelectionChangedCause.toolbar,
                      );
                    },
                    child: const Icon(Icons.copy_all),
                  ),
                ],
              );
            },
            onSelectionChanged: (value) {
              debugPrint('onSelectionChanged: ${value?.plainText}');
            },
```

Video:


https://github.com/flutter/flutter/assets/153862654/cf9740e5-4d27-4253-a678-bbac55d45696



I'm doing something bad?

**Comment by Renzo-Olivares** [comment]
@vergaraSC What platform are you on?

**Comment by vergaraSC** [comment]
> @vergaraSC What platform are you on?

The platform is web-based, kIsWeb. 

Or do you mean something else?

**Comment by Renzo-Olivares** [comment]
@vergaraSC I mean the underlying native platform that the web is running on like `Android`.

**Comment by vergaraSC** [comment]
> @vergaraSC I mean the underlying native platform that the web is running on like `Android`.

oh sorry, yeah the platform is **macOS**

**Comment by Renzo-Olivares** [comment]
@vergaraSC I just tried it on the `master` branch and it works for me using `macOS` + `chrome`. The example https://api.flutter.dev/flutter/material/SelectionArea-class.html also works for me on stable. Do you have any inline widget spans in your text?

**Comment by vergaraSC** [comment]
> @vergaraSC I just tried it on the `master` branch and it works for me using `macOS` + `chrome`. The example https://api.flutter.dev/flutter/material/SelectionArea-class.html also works for me on stable. Do you have any inline widget spans in your text?

Yes, I have Text.rich and in its children I have InlineSpan widgets, how could I solve this problem? should I look for another approach?

**Comment by Renzo-Olivares** [comment]
@vergaraSC For now there is nothing you can do to solve this except remove `InlineSpan` widgets. The reason is likely that the `InlineSpan` widgets are throwing off the selectable ordering. I am working on a fix for this.

**Comment by vergaraSC** [comment]
> @vergaraSC For now there is nothing you can do to solve this except remove `InlineSpan` widgets. The reason is likely that the `InlineSpan` widgets are throwing off the selectable ordering. I am working on a fix for this.

I understand, thank you very much for the clarification and great that you are working on a potential fix.
